### PR TITLE
Asserter: New `resource` asserter!

### PR DESCRIPTION
**Pull request by Hywan**
Example: `$this->resource($variable)->isOfType('stream')`.

### CODE REVIEW

**Comment by mikaelrandy** [comment]
Huge :+1: on the idea ! I have often thought i need this kind of asserter ;)

And :+1: on the code

File: classes/asserters/phpResource.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		}
+		else
+		{
+			$this->fail($this->_('%s is not a resource', $this));
+		}
+
+		return $this;
+	}
+
+	public function isOfType($type, $failMessage = null)
```
Comment (jubianchi) [inline classes/asserters/phpResource.php]: I would add some helper assertions here like `isStream`, `isMysql`, ... and a wildcard helper via `__get/__call` (`is*`) to allow things like `->isDomXml`, `->isWhatever`, ...

File: classes/test/adapter/calls.php [hunk 2]
```diff
@@ -320,6 +320,10 @@ protected function setCall(adapter\call $call, $position = null)
 			throw new exceptions\logic\invalidArgument('Function is undefined');
 		}
 
+		if (!isset(self::$callsNumber)) {
```
Comment (jubianchi) [inline classes/test/adapter/calls.php]: Why ?

This seems a bit unrelated ?

File: classes/asserters/phpResource.php (hunk 1 above)
Comment (Hywan) [inline classes/asserters/phpResource.php]: Maybe we should all native resource, but not the other, like MySQL.

File: classes/tools/variable/analyzer.php [hunk 3]
```diff
@@ -83,6 +83,11 @@ public function isArray($mixed)
 		return (is_array($mixed) === true);
 	}
 
+	public function isResource($mixed)
+	{
+		return (is_resource($mixed) === true);
```
Comment (jubianchi) [inline classes/tools/variable/analyzer.php]: could you also add a `case` in `getTypeOf`please

File: classes/asserters/phpResource.php (hunk 1 above)
Comment (Hywan) [inline classes/asserters/phpResource.php]: http://php.net/resource hmmm the list is very very long :-p.

File: classes/test/adapter/calls.php (hunk 2 above)
Comment (Hywan) [inline classes/test/adapter/calls.php]: It is, or not. If you are testing a resource that of kind `atoum://` stream, you will have an error. See https://github.com/Hywan/atoum/commit/d898c6d6111e98a38555a6ba93b2513198aa3007 for more “explanations”.

File: classes/tools/variable/analyzer.php (hunk 3 above)
Comment (Hywan) [inline classes/tools/variable/analyzer.php]: Already present.

Comment (Hywan) [inline classes/tools/variable/analyzer.php]: (Go check https://github.com/atoum/atoum/pull/535#discussion-diff-47633946R26).

File: tests/units/classes/asserters/phpResource.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		$this
+			->given($this->newTestedInstance)
+			->then
+				->object($this->testedInstance->getGenerator())->isEqualTo(new asserter\generator())
+				->object($this->testedInstance->getAnalyzer())->isEqualTo(new variable\analyzer())
+				->object($this->testedInstance->getLocale())->isEqualTo(new atoum\locale())
+				->variable($this->testedInstance->getValue())->isNull()
+				->boolean($this->testedInstance->wasSet())->isFalse()
+
+			->given($this->newTestedInstance($generator = new asserter\generator(), $analyzer = new variable\analyzer(), $locale = new atoum\locale()))
```
Comment (jubianchi) [inline tests/units/classes/asserters/phpResource.php]: why not using the newly introduced auto-mocks ?

```php
public function __construct(asserter\generator $generator, variable\analyzer $analyzer, atoum\locale $locale)
{
    $this
        ->given($this->newTestedInstance($generator, $analyzer, $locale)
        ->then
            ->object($this->testedInstance->getGenerator())-> isIdenticalTo($generator)
            //...
    ;
}
```

File: tests/units/classes/asserters/phpResource.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+				->boolean($this->testedInstance->wasSet())->isFalse()
+		;
+	}
+
+	public function testSetWith()
+	{
+		$this
+			->given(
+				$asserter = $this->newTestedInstance
+					->setAnalyzer($analyzer = new \mock\atoum\tools\variable\analyzer())
```
Comment (jubianchi) [inline tests/units/classes/asserters/phpResource.php]: this mock is not used

File: tests/units/classes/asserters/phpResource.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+					->setLocale($locale = new \mock\atoum\locale()),
+				$this->calling($locale)->_ = $notAResource = uniqid()
+			)
+			->then
+				->exception(function() use ($asserter, & $value) { $asserter->setWith($value = uniqid()); })
+					->isInstanceOf('mageekguy\atoum\asserter\exception')
+					->hasMessage($notAResource)
+				->mock($locale)->call('_')->withArguments('%s is not a resource', $asserter)->once
```
Comment (jubianchi) [inline tests/units/classes/asserters/phpResource.php]: can you indent the `->call` assertion please

File: tests/units/classes/asserters/phpResource.php [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+				->string($asserter->getValue())->isEqualTo($value)
+
+				->object($asserter->setWith($value = fopen(__FILE__, 'r')))->isIdenticalTo($asserter)
```
Comment (jubianchi) [inline tests/units/classes/asserters/phpResource.php]: did you try using the VFS here ?

File: classes/asserters/phpResource.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	{
+		$actualType = get_resource_type($this->valueIsSet()->value);
+
+		if ($actualType === $type)
+		{
+			$this->pass();
+		}
+		else
+		{
+			$this->fail(($failMessage ?: $this->_('%s is not of type %s', $this, $type)) . PHP_EOL . $this->diff($actualType));
```
Comment (jubianchi) [inline classes/asserters/phpResource.php]: I don't think a diff is useful here for a simple string comparison

File: tests/units/classes/asserters/phpResource.php (hunk 4 above)
Comment (Hywan) [inline tests/units/classes/asserters/phpResource.php]: Because I just copy/paste test cases from the `integer` class ;-).

Comment (Hywan) [inline tests/units/classes/asserters/phpResource.php]: And your proposal does not work :-).

File: tests/units/classes/asserters/phpResource.php (hunk 5 above)
Comment (Hywan) [inline tests/units/classes/asserters/phpResource.php]: :+1:

File: tests/units/classes/asserters/phpResource.php (hunk 6 above)
Comment (Hywan) [inline tests/units/classes/asserters/phpResource.php]: Again, I just copied/pasted from the `integer` test suite ;-). Fixed :+1:

File: tests/units/classes/asserters/phpResource.php (hunk 7 above)
Comment (Hywan) [inline tests/units/classes/asserters/phpResource.php]: Much heavier than using `fopen(__FILE__)`, and the result is exactly the same :-).

File: classes/asserters/phpResource.php (hunk 8 above)
Comment (Hywan) [inline classes/asserters/phpResource.php]: What do you suggest? Removing the diff?

File: classes/asserters/phpResource.php (hunk 1 above)
Comment (jubianchi) [inline classes/asserters/phpResource.php]: this is why I suggested a wildcard helper ;)

File: classes/tools/variable/analyzer.php (hunk 3 above)
Comment (jubianchi) [inline classes/tools/variable/analyzer.php]: my bad :)

File: classes/asserters/phpResource.php (hunk 8 above)
Comment (jubianchi) [inline classes/asserters/phpResource.php]: yes

File: tests/units/classes/asserters/phpResource.php (hunk 4 above)
Comment (jubianchi) [inline tests/units/classes/asserters/phpResource.php]: it works here: https://github.com/atoum/atoum/pull/537/files#diff-f1d31228248836643d3c15038bcb5e73R76

File: classes/asserters/phpResource.php (hunk 1 above)
Comment (Hywan) [inline classes/asserters/phpResource.php]: What to do with space or hyphen? For instance “persistent cubrid connection” is transformed into…?

File: tests/units/classes/asserters/phpResource.php (hunk 4 above)
Comment (jubianchi) [inline tests/units/classes/asserters/phpResource.php]: ok I forgot a `)` in my snippet but it works like a charm :p

File: classes/asserters/phpResource.php (hunk 1 above)
Comment (Hywan) [inline classes/asserters/phpResource.php]: Same question with “dbx_result_object”.

Comment (Hywan) [inline classes/asserters/phpResource.php]: Another question with “gd PS font”.

Comment (Hywan) [inline classes/asserters/phpResource.php]: Because we should transform `isPersistentCubridConnection` to `persistent cubric connection`, fine.
But `isDbx_result_object`… crazy.
And `isGdPSFont` will be transformed into `gd ps font`, not `gd PS font`.

:-/

Comment (jubianchi) [inline classes/asserters/phpResource.php]: We should compare both values in lowercase. But I agree that we won't be able to cover 100% with helper like `isMysql`or `isStream` or the wildcard helper. This is why we still need the `isOfType` asssertion.

File: classes/test/adapter/calls.php [hunk 9]
```diff
@@ -320,6 +320,12 @@ protected function setCall(adapter\call $call, $position = null)
 			throw new exceptions\logic\invalidArgument('Function is undefined');
 		}
 
+		// A bug in PHP removes this static attribute when destructing all the
+		// object. So accessing it create an error. This is a workaround.
+		if (!isset(self::$callsNumber)) {
```
Comment (jubianchi) [inline classes/test/adapter/calls.php]: https://github.com/Hywan/atoum/pull/4

File: tests/units/classes/asserters/phpResource.php (hunk 4 above)
Comment (Hywan) [inline tests/units/classes/asserters/phpResource.php]: @jubianchi Can you make a PR, I don't understand your purpose (and I wasn't speaking about the syntax error ;-)).

File: classes/asserters/phpResource.php (hunk 1 above)
Comment (Hywan) [inline classes/asserters/phpResource.php]: Fixed.

Comment (Hywan) [inline classes/asserters/phpResource.php]: I have found a trick to support all the possible kind of resource types.

File: classes/test/adapter/calls.php (hunk 9 above)
Comment (Hywan) [inline classes/test/adapter/calls.php]: It seems to break the tests :-/.

File: tests/units/classes/asserters/phpResource.php [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+						->withArguments('%s is not a resource', $asserter)
+						->once
+				->string($asserter->getValue())->isEqualTo($value)
+
+				->object($asserter->setWith($value = fopen(atoum\mock\streams\fs\file::get(), 'r')))->isIdenticalTo($asserter)
+				->resource($asserter->getValue())->isEqualTo($value)
+		;
+	}
+
+	public function testIsOfType()
```
Comment (jubianchi) [inline tests/units/classes/asserters/phpResource.php]: this test should be renamed to `testMatches`

**Comment by jubianchi** [comment]
@Hywan with your last commits, failure message are a bit weird. For example: `Resource id #90 of type stream is not of type .?foo.?bar`

I think it's better to not rely on pattern matching at all and only handle common cases (eg split camel case and transform to space-separated string).

Edge-cases would be handled by the user using is `isOfType` assertion.

What do you think ?

File: tests/units/classes/asserters/phpResource.php (hunk 10 above)
Comment (Hywan) [inline tests/units/classes/asserters/phpResource.php]: Done.

**Comment by Hywan** [comment]
@jubianchi Here is what I did: `isOfType` with a regular string comparison, and `matches` with regular expression, and `is*` asserters fallback to `matches`. This is pretty logical this way.

**Comment by Hywan** [comment]
Everything seems good. Waiting on tests.

**Comment by Hywan** [comment]
Hmm, still a segfault…

**Comment by Hywan** [comment]
@jubianchi The segfault was caused by using VFS instead of a plain `fopen(__FILE__, 'r')` ;-).

**Comment by Hywan** [comment]
Ready to merge! Here is final API:
```php
$this
    ->resource($fp)
        ->isOfType('stream')
        ->isIdenticalTo($anotherFp);
```
(Because we inherit from `variable` we can use `isIdenticalTo` & co.).
```php
$this
    ->resource($fp)
        ->isOfType('stream')
        ->matches('stream')
        ->isStream()
```
All these examples are similar.
However, is `resource::matches` has the same behavior than `string::matches`? No. `string::matches` expect a full PCRE expression, like `/stream/` for instance, while here it is just `stream`. This is not consistent. I will update that.

**Comment by Hywan** [comment]
The final API:
Ready to merge! Here is the current API:
```php
$this
    ->resource($fp)
        ->isOfType('stream')
        ->isIdenticalTo($anotherFp)
        ->matches('/^stream$/')
        ->isStream()
```

**Comment by Hywan** [comment]
Readyyy :-].

**Comment by jubianchi** [comment]
I think the API is a bit misleading, especially, the `matches` assertion does not seem appropriate to me (not clear enough).

User might expect that atoum will try to match the pattern against the stream contents.

To assert on stream type, what do you think of implementing something like:

```php
<?php

$this
    ->resource($fp)
        ->isOfType('stream')
        ->isStream()
        ->type
            ->matches('/^stream$/')
```

All assertions here are equivalent, they will assert on the stream type. The only difference with your proposal is the `->type->...` form: we fallback to a `string` asserter to assert on the type value letting us do more assertions (`isEqualTo`, `matches`, `startsWith`, ...)

**Comment by Hywan** [comment]
@jubianchi How to implement this without breaking the “assertion flow”? For instance `->resource(…)->type->isEqualTo(…)->array(…)->…`?

**Comment by jubianchi** [comment]
@Hywan see:

* https://github.com/atoum/atoum/blob/master/classes/asserters/exception.php#L22-L23
* https://github.com/atoum/atoum/blob/master/classes/asserters/exception.php#L149-L152

;)

File: classes/asserters/phpResource.php [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+		{
+			$this->fail($failMessage ?: $this->_('%s does not match %s', $this, $pattern));
+		}
+
+		return $this;
+	}
+
+	public function __call($name, $arguments)
+	{
+		if ('is' === substr($name, 0, 2)) {
```
Comment (jubianchi) [inline classes/asserters/phpResource.php]: `{` should be on its own line

File: classes/asserters/phpResource.php [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+			$pattern = preg_replace_callback(
+				'/([A-Z])([a-z]+)/',
+				function ($matches) {
+					return '.?' . strtolower($matches[1]) . $matches[2];
+				},
+				$pattern
+			);
+			$pattern = '/^' . $pattern . '$/i';
+
+			if (1 === count($arguments)) {
```
Comment (jubianchi) [inline classes/asserters/phpResource.php]: `{` should be on its own line

File: classes/asserters/phpResource.php [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+				return $this->matches($pattern, $arguments[0]);
+			} else {
```
Comment (jubianchi) [inline classes/asserters/phpResource.php]: `}` and `{` should be on their own line

**Comment by jubianchi** [comment]
@Hywan do you think you can make your PR ready before the 2.5.0 due date (01-07-2016) ?

If no, do you allow me to apply the required refactoring to your branch and merge it ?

**Comment by Hywan** [comment]
@jubianchi I will try for the next 30mn ;-).

**Comment by Hywan** [comment]
@jubianchi Normally it's done and it works. The test fails because “the branch does not exist”, only for some version of PHP. I guess this is an infrastructure issue, and not related to atoum or this PR. The last commit was necessary and it will run a new fresh test campaign. Wait and see.

**Comment by Hywan** [comment]
Green :white_check_mark: \o/.

**Comment by jubianchi** [comment]
merge in 601663a72fafe9b
